### ISSUE TITLE
Realtime indexing error on Arbitrum 

### ISSUE THREAD

**Issue by kev-daniell** [ponder-sh/ponder#546]
i tried using ponder recently for indexing transactions that have been settled on our contract. currently got it to work well, historical indexing works perfectly for both mainnet & arb.

however realtime indexing seems to be failing after 20 or so seconds for arbitrum (mainnet seems ok). attached are my config and the error msg.
![telegram-cloud-photo-size-1-4909177340957928222-y](https://github.com/ponder-sh/ponder/assets/83150865/3581b01c-d0af-4dd0-ba3e-5dd62a4df5d7)

![telegram-cloud-photo-size-1-4909177340957928223-y](https://github.com/ponder-sh/ponder/assets/83150865/6a5b16e3-17fb-4748-8e91-54a3c6d976ea)

**Comment by kyscott18** [comment]
Can you share some information about what your indexing functions are doing? Any contract reads?

**Comment by kev-daniell** [comment]
sure thing, apologies for the sort of messy code as i was trying to get ponder working as a proof of concept: 

tldr: we get transaction hash and other txn info from the event and essentially use that information to update records in our offchain mongodb for analytic purposes 

```
import { ponder } from "@/generated";
import "dotenv/config";
import { MongoClient } from "mongodb";
import { Seaport__factory } from "../types/ethers-contracts";
import { MongoCollections, MongoDatabases } from "./constants";
import { clients } from "./providers";
import { tokenMap } from "./tokens";

export const mongoClient = new MongoClient(process.env.MONGO_DB_URI || "");

ponder.on("AoriProtocol:TradeOccurred", async ({ event, context }: any) => {
  // @ts-ignore
  const receipt = await clients[context.network.chainId].getTransactionReceipt(event.transaction.hash);

  if (!receipt) return;

  receipt.logs.map(async (log: any) => {
    const res = Seaport__factory.createInterface().parseLog(log as any)
    if (res?.name !== "OrdersMatched") return;

    const orderHash = res.args[0][0];
    const txHash = event.transaction.hash;

    const order = await mongoClient
      .db(MongoDatabases.OrderManager)
      .collection(MongoCollections.Orders)
      .findOne({ orderHash })
    if (!order) {
      console.log("no order")
      return
    }

    if (order.fulfilledAt != undefined) {
      console.log("order already fulfilled")
      return
    }

    const taker = order.pendingTakerOrders[order.pendingTakerOrders.length - 1].parameters.offerer

    const prices = await mongoClient
      .db(MongoDatabases.DataIndexers)
      .collection(MongoCollections.AssetPricing)
      .find({ $or: [{ address: order.inputToken }, { address: order.outputToken }] })
      .toArray()

    const inputTokenInfo = prices.find((x) => x.address === order.inputToken);
    const outputTokenInfo = prices.find((x) => x.address === order.outputToken);

    if (!inputTokenInfo || !outputTokenInfo) throw new Error("Cannot fetch asset pricse in Order Fulfillment");

    const inputToken = tokenMap.get(order.inputToken)
    const outputToken = tokenMap.get(order.outputToken)

    if (!inputToken || !outputToken) {
      console.log("no token info: ", order.inputToken, order.outputToken)
      return
    }

    const inputAmount = inputTokenInfo.price * (order.totalInput / (10 ** inputToken.decimals))
    const outputAmount = outputTokenInfo.price * (order.totalOutput / (10 ** outputToken.decimals))

    const result = await mongoClient
      .db(MongoDatabases.OrderManager)
      .collection(MongoCollections.Orders)
      .updateOne({
        orderHash
      }, {
        $set: {
          totalInputUSD: inputAmount,
          totalOutputUSD: outputAmount,
          taker,
          fulfilledAt: Date.now(),
          txnHash: txHash
        },
      });

    console.log("(Order Fulfilled) orderHash: ", res.args[0][0], "txnHash: ", event.transaction.hash);
  })
});
```

**Comment by kyscott18** [comment]
I think I see what is going wrong here. I don't know the exact numbers but let's say the call you do to getTransactionReceipt() takes 100 milliseconds. Arbitrum tries to emit 15 blocks per second (67 milliseconds per block). So even without any other operations in the indexing function, there is no way to keep up with the chain. 

The solve for this would be to try and add the seaport contract to the config rather than fetch the logs within a separate indexing function. 

For example your config could look like

```ts
export default createConfig({
  networks: {
    mainnet: {
      chainId: 1,
      transport: http(process.env.PONDER_RPC_URL_1),
    },
    arbitrum: {
      chainId: 42161,
      transport: http(process.env.PONDER_RPC_URL_42161),
    },
  },
  contracts: {
    Aori: {
      abi: <aori abi>,
      network: {
        mainnet: {
          address: <mainnet address>,
        },
        arbitrum: {
          address: <arbitrum address>
        },
      },
    },
    Seaport: {
      abi: <seaport abi>,
      network: {
        mainnet: {
          address: <mainnet address>,
        },
        arbitrum: {
          address: <arbitrum address>
        },
      },
      filter: {
        event: "OrdersMatched",
      },
    },
  },
});
```

Now you have to write indexing function for both `Seaport:OrdersMatched` and `Aori:TradeOcurred`. If you know the ordering of the events before hand, you could add some logic like:

```ts
let aoriTx: Hex | undefined = undefined;

ponder.on("Aori:TradeOccured", async ({ event, context }) => {
  aoriTx = event.transaction.hash;
});

ponder.on("Seaport:OrdersMatched", async ({event, context }) => {
  if (aoriTx === event.transaction.hash) {
    // Now you know that this event occurred in same transaction as the Aori event
    // Do your mangoDB logic here
  }
});
```

Hopefully this makes sense. Happy to help more
